h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -141,6 +142,87 @@ public void computeCrossValidation() {
     super.computeCrossValidation();
   }
 
+
+  /* This method aligns the submodels across CV folds.
+   *
+   * In the following depiction each submodel is represented by a tuple (alpha_value, lambda_value). Ideally each CV fold
+   * should have the same submodels but there can be differences due to max_iteration and max_runtime_secs constraint.
+   * Iterations are reset when alpha value change which can cause missing model in the middle and max_runtime_secs constraint
+   * can cause missing submodels at the end of the submodel array.
+   *
+   *  CV1:
+   *   +------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *   | 0, 1 | 0, 0.9 | 0, 0.8 | 0, 0.7 | 0, 0.6 | 1, 1 | 1, 0.9 | 1, 0.8 | 1, 0.7 | 1, 0.6 | 1, 0.5 | 1, 0.4 |
+   *   +------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *  CV2:
+   *   +------+--------+--------+--------+--------+--------+------+--------+--------+--------+
+   *   | 0, 1 | 0, 0.9 | 0, 0.8 | 0, 0.7 | 0, 0.6 | 0, 0.5 | 1, 1 | 1, 0.9 | 1, 0.8 | 1, 0.7 |
+   *   +------+--------+--------+--------+--------+--------+------+--------+--------+--------+
+   *
+   *    ||
+   *    ||
+   *   \  /
+   *    \/
+   *
+   *  Aligned Submodels:
+   *
+   *  CV1:
+   *  +------+--------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *  | 0, 1 | 0, 0.9 | 0, 0.8 | 0, 0.7 | 0, 0.6 |  NULL  | 1, 1 | 1, 0.9 | 1, 0.8 | 1, 0.7 | 1, 0.6 | 1, 0.5 | 1, 0.4 |
+   *  +------+--------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *
+   *  CV2:
+   *  +------+--------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *  | 0, 1 | 0, 0.9 | 0, 0.8 | 0, 0.7 | 0, 0.6 | 0, 0.5 | 1, 1 | 1, 0.9 | 1, 0.8 | 1, 0.7 |  NULL  |  NULL  |  NULL  |
+   *  +------+--------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *
+   */
+  private double[] alignSubModelsAcrossCVModels(ModelBuilder[] cvModelBuilders) {
+    // maximum index of alpha change across all the folds
+    int[] alphaChangePoints = new int[_parms._alpha.length + 1];
+    int[] alphaSubmodels = new int[_parms._alpha.length];
+    for (int i = 0; i < cvModelBuilders.length; ++i) {
+      GLM g = (GLM) cvModelBuilders[i];
```

Comment:
When GLM concludes with best_submodel_idx, it chooses among all the alpha and lambda values, what is the best combo.  So, instead of searching through all alpha and lambda combos again for CV, look at the best_submodel_idx submodel, look at the submodel, take that alpha value to use instead of having to search through all.  Then collect the lambdas associated with that alpha value.  Please see the pictures in my comments.  If between multiple runs, you do see different alphas that is the best submodel, you then will add that alpha to the list.  In addition, when you pick the lambdas after deciding on an alpha, you only need to take the lambdas that the best_submodel_idx dictates.  No need to take all the lambdas in since overfitting is happening with the later submodel lambdas

Reply:
First of all, thank you @wendycwong  for the review.

Intuitively, I'd say that it might lead to a suboptimal choice of `alpha` and `lambda`. I don't have any proof of it but it vaguely reminds me of the [Stein's paradox](https://en.wikipedia.org/wiki/Stein%27s_example), in other words looking at local optima obtained from the individual folds might not lead to picking the global optimum for all the folds. 

From the performance point of view, the search through all the `alpha`/`lambda` combos should be quick if there won't be any call to the `computeSubmodel` - basically we go through all the submodels and compute the mean of already computed metrics. 

The slowest part of the code is the part that computes the submodel which is in this PR already limited only to the missing submodels (which provides significant speedup on [one dataset](https://www.openml.org/d/41167) that I used to test it (from ~70mins to 20mins)).

To really speed it up further I'd propose either not computing any submodels in the `cv_computeAndSetOptimalParameters` or computing only those submodels that are best in some of the folds. But this kind of change might lead to worse predictive performance so maybe we should do this only if the user specified some nonzero value to `_max_runtime_secs`. 

Note that I am not a GLM expert and my intuition might be completely wrong, I just want to make sure we won't make a worse estimator for some (possibly negligible) performance improvement.

---

Just to make sure should the following equalities be always satisfied `_selected_submodel_idx == _best_submodel_idx == _best_lambda_idx`?

Looking into picking the alpha and lambda values, I noticed that we probably don't set the `_selected_(alpha|lambda)_idx`  correctly - it seems to assume that the glm will create submodels for all the lambdas which is not what happens in my experience. https://github.com/h2oai/h2o-3/blob/d148b0ef279502c3aa074e273173fd9755bc3635/h2o-algos/src/main/java/hex/glm/GLMModel.java#L1460-L1463